log4j-api/src/main/java/org/apache/logging/log4j/internal/map/UnmodifiableArrayBackedMap.java

```diff
@@ -350,82 +351,59 @@ public UnmodifiableArrayBackedMap copyAndRemove(String key) {
     }
 
     /**
-     * Creates a new instance that contains the same entries as this map, minus all
-     * of the keys passed in the arguments.
-     *
-     * @param key
-     * @param value
-     * @return
+     * Creates a new instance where the entries of provided keys are removed.
      */
     public UnmodifiableArrayBackedMap copyAndRemoveAll(Iterable<String> keysToRemoveIterable) {
+
+        // Short-circuit if the map is empty
         if (isEmpty()) {
-            // shortcut: if this map is empty, the result will continue to be empty
             return EMPTY_MAP;
         }
 
-        // now we build a Set of keys to remove
-        Set<String> keysToRemoveSet;
-        if (keysToRemoveIterable instanceof Set) {
-            // we already have a set, let's cast it and reuse it
-            keysToRemoveSet = (Set<String>) keysToRemoveIterable;
-        } else {
-            // iterate through the keys and build a set
-            keysToRemoveSet = new HashSet<>();
-            for (String key : keysToRemoveIterable) {
-                keysToRemoveSet.add(key);
-            }
-        }
+        // Collect distinct keys to remove
+        final Set<String> keysToRemove = keysToRemoveIterable instanceof Set
+                ? (Set<String>) keysToRemoveIterable
+                : StreamSupport.stream(keysToRemoveIterable.spliterator(), false)
```

Comment:
Streams create intermediate objects and add instructions.  Not a great idea in a performance-sensitive spot.

Reply:
Fixed in 91ff32cec924a06611608bb8d86393c3e1c4be4e.